rllib/utils/exploration/parameter_noise.py

```diff
@@ -0,0 +1,386 @@
+        # This is due to the fact that the optimizer does it's step right
+        # after the episode was reset (and hence the noise was already added!).
+        # We don't want to update into a noisy net.
+        self.episode_started = True
+
+    def _delayed_on_episode_start(self, tf_sess):
+        # Sample fresh noise and add to weights.
+        if self.default_explore:
+            self._sample_new_noise_and_add(tf_sess=tf_sess, override=True)
+        # Only sample, don't apply anything to the weights.
+        else:
+            self._sample_new_noise(tf_sess=tf_sess)
+        self.episode_started = False
+
+    @override(Exploration)
+    def on_episode_end(self,
+                       policy,
+                       *,
+                       environment=None,
+                       episode=None,
+                       tf_sess=None):
+        # Remove stored noise from weights (only if currently noisy).
+        if self.weights_are_currently_noisy:
+            self._remove_noise(tf_sess=tf_sess)
+
+    @override(Exploration)
+    def postprocess_trajectory(self, policy, sample_batch, tf_sess=None):
+        noisy_action_dist = noise_free_action_dist = None
+        # Adjust the stddev depending on the action (pi)-distance.
+        # Also see [1] for details.
+        inputs, dist_class, _ = policy.compute_distribution_inputs(
+            obs_batch=sample_batch[SampleBatch.CUR_OBS],
+            # TODO(sven): What about state-ins and seq-lens?
+            prev_action_batch=sample_batch.get(SampleBatch.PREV_ACTIONS),
+            prev_reward_batch=sample_batch.get(SampleBatch.PREV_REWARDS),
+            explore=self.weights_are_currently_noisy)
+
+        # Categorical case (e.g. DQN).
+        if dist_class is Categorical:
+            action_dist = softmax(inputs)
+        else:  # TODO(sven): Other action-dist cases.
+            raise NotImplementedError
+
+        if self.weights_are_currently_noisy:
+            noisy_action_dist = action_dist
+        else:
+            noise_free_action_dist = action_dist
+
+        inputs, dist_class, _ = policy.compute_distribution_inputs(
+            obs_batch=sample_batch[SampleBatch.CUR_OBS],
+            # TODO(sven): What about state-ins and seq-lens?
+            prev_action_batch=sample_batch.get(SampleBatch.PREV_ACTIONS),
+            prev_reward_batch=sample_batch.get(SampleBatch.PREV_REWARDS),
+            explore=not self.weights_are_currently_noisy)
+
+        # Categorical case (e.g. DQN).
+        if dist_class is Categorical:
+            action_dist = softmax(inputs)
+
+        if not self.weights_are_currently_noisy:
+            noisy_action_dist = action_dist
+        else:
+            noise_free_action_dist = action_dist
+
+        # Categorical case (e.g. DQN).
+        if dist_class is Categorical:
+            # Calculate KL-divergence (DKL(clean||noisy)) according to [2].
+            kl_divergence = np.nanmean(
```

Comment:
Why not use `action_dist.kl(other_action_dist)`?

I think these computations should be abstracted away in the action dist class itself, so that we don't need to peek at the "type" of the distribution and do different logic depending on that.

Reply:
It's happening off-graph. Our distributions are not tested for framework=None (have no numpy implementations).

Reply:
I see, is this a long term TODO then? It seems that ideally we would compute the kl using the action dist objects.

Reply:
Actually, why not do this on-graph?

Reply:
Absolutely! But yeah, thus far, all distributions, torch/tf, were always in graph, so we cannot use those unless we add framework=None support to all of them.

Reply:
Ok, can you leave a TODO here so the reader knows why the code is currently in this form?

Reply:
Done.